src/anemoi/inference/runner.py

```diff
@@ -391,186 +235,121 @@ def run(
+            output = np.squeeze(y_pred.cpu().numpy())  # shape: (values, variables)
 
-            prognostic_fields_numpy = output[:, prognostic_output_mask]
-            if len(diagnostic_output_mask):
-                diagnostic_fields_numpy = output[:, diagnostic_output_mask]
+            # Update state
+            for i in range(output.shape[1]):
+                result["fields"][self.checkpoint.output_tensor_index_to_variable[i]] = output[:, i]
 
-            for n, (m, param) in enumerate(zip(prognostic_data_from_retrieved_fields_mask, prognostic_params)):
-                template = reference_fields[m]
-                assert template.datetime()["valid_time"] == most_recent_datetime, (
-                    template.datetime()["valid_time"],
-                    most_recent_datetime,
-                )
-                output_callback(
-                    prognostic_fields_numpy[:, n],
-                    template=template,
-                    step=step,
-                    check_nans=True,  # param in can_be_missing,
-                )
+            yield result
 
-            # Write diagnostic fields
-            if len(diagnostic_output_mask):
-                for n, param in enumerate(self.checkpoint.diagnostic_params):
-                    accumulated_output[n] += np.maximum(0, diagnostic_fields_numpy[:, n])
-                    assert prognostic_template.datetime()["valid_time"] == most_recent_datetime, (
-                        prognostic_template.datetime()["valid_time"],
-                        most_recent_datetime,
-                    )
-
-                    if param in accumulations_params:
-                        output_callback(
-                            accumulated_output[n],
-                            stepType="accum",
-                            template=prognostic_template,
-                            startStep=0,
-                            endStep=step,
-                            param=param,
-                            check_nans=True,  # param in can_be_missing,
-                        )
-                    else:
-                        output_callback(
-                            diagnostic_fields_numpy[:, n],
-                            template=prognostic_template,
-                            step=step,
-                            check_nans=True,  # param in can_be_missing,
-                        )
-
-            # Next step
-
-            prognostic_fields = y_pred[..., prognostic_output_mask]
-
-            # Compute new forcing
-
-            forcing = forcing_and_constants(
-                source=input_fields[:1],
-                param=self.checkpoint.computed_forcings,
-                date=start_datetime + datetime.timedelta(hours=step),
-            )
-            forcing = np.swapaxes(forcing[np.newaxis, np.newaxis, ...], -2, -1)
-            forcing = torch.from_numpy(forcing).to(device)
+            # Update  tensor for next iteration
```

Comment:
I think due to the different ways to prepare for the next iteration, this functionality should be abstracted out.

Reply:
Let's merge and refactor in another PR